### PR TITLE
Fix(RuleRight): Take db groups into account for rights rules on LDAusers

### PR DESCRIPTION
**Pull request by stonebuzz**
same as https://github.com/glpi-project/glpi/pull/16548 but for main

The "groups" criteria for authorizations rules is not processed correctly for LDAP users.

Small improvement added to the unit test ```RuleBuilder``` to manages the ```subtype```, rather than the ```CreateRule``` function.

| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | !31311

### CODE REVIEW

File: src/User.php [hunk 1]
```diff
@@ -2096,6 +2117,17 @@ public function getFromLDAP($ldap_connection, array $ldap_method, $userdn, $logi
...
+                $searched_user = new User();
+                if ($searched_user->getFromDBbyDnAndAuth($userdn, $ldap_method["id"])) {
+                    $groups = array_merge($groups, array_column(Group_User::getUserGroups($searched_user->getID()), 'id'));
+                }
+
+                if ($this->fields['name'] == "brazil7") {
+                    var_dump($groups);
+                }
+
+
```
Comment (cedric-anne) [inline src/User.php:2130]:
```suggestion
```

File: tests/LDAP/AuthLdap.php [hunk 2]
```diff
@@ -2367,6 +2369,57 @@ public function testRuleRight()
...
+        // Add group to user
+        $this->createItem(Group_User::class, [
+            'groups_id' => $group_id,
+            'users_id'  => $users_id,
+        ]);
+        $groups = (new Group_User())->find([
+            'groups_id' => $group_id,
+            'users_id'  => $users_id,
+        ]);
+        $this->array($groups)->hasSize(1);
```
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php:2410]: This is useless as `this->createItem()` already checks that the group exists.

File: tests/LDAP/AuthLdap.php [hunk 3]
```diff
@@ -2367,6 +2369,57 @@ public function testRuleRight()
...
+        $builder->addCriteria('_groups_id', \Rule::PATTERN_IS, $group_id);
+        $builder->addAction('assign', 'profiles_id', 4); // Super admin
+        $builder->addAction('assign', 'entities_id', 1);
+        $builder->setEntity(0);
+        $this->createRule($builder);
+
+        // login the user to force a real synchronisation (and creation into DB)
+        $this->login('brazil7', 'password', false);
+        $users_id = \User::getIdByName('brazil7');
+
```
Comment (cedric-anne) [inline tests/LDAP/AuthLdap.php:2400]: You should ensure that the user is not attached to the profile at creation.

```suggestion

        // Check that the user is not attached to the profile at creation
        $rights = (new Profile_User())->find([
            'users_id' => $users_id,
            'profiles_id' => 4,
        ]);

        $this->array($rights)->hasSize(0);

```

**Review by cedric-anne — CHANGES_REQUESTED** [review]

**Review by AdrienClairembault — APPROVED** [review]

File: tests/LDAP/AuthLdap.php (hunk 2 above)
Comment (stonebuzz) [reply tests/LDAP/AuthLdap.php]: fixed

File: tests/LDAP/AuthLdap.php (hunk 3 above)
Comment (stonebuzz) [reply tests/LDAP/AuthLdap.php:2400]: fixed

**Review by cedric-anne — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**